freqtrade/rpc/emc.py

```diff
@@ -0,0 +1,234 @@
+        # callbacks for the messages
+        self.topics = [RPCMessageType.WHITELIST, RPCMessageType.ANALYZED_DF]
+
+        self.start()
+
+    def start(self):
+        """
+        Start the main internal loop in another thread to run coroutines
+        """
+        self._loop = asyncio.new_event_loop()
+
+        if not self._thread:
+            logger.info("Starting ExternalMessageConsumer")
+
+            self._thread = Thread(target=self._loop.run_forever)
+            self._thread.start()
+            self._running = True
+        else:
+            raise RuntimeError("A loop is already running")
+
+        self._main_task = asyncio.run_coroutine_threadsafe(self._main(), loop=self._loop)
+
+    def shutdown(self):
+        """
+        Shutdown the loop, thread, and tasks
+        """
+        if self._thread and self._loop:
+            logger.info("Stopping ExternalMessageConsumer")
+
+            if self._sub_tasks:
+                # Cancel sub tasks
+                for task in self._sub_tasks:
+                    task.cancel()
+
+            if self._main_task:
+                # Cancel the main task
+                self._main_task.cancel()
+
+            self._thread.join()
+
+    async def _main(self):
+        """
+        The main task coroutine
+        """
+        rpc_lock = asyncio.Lock()
+
+        try:
+            # Create a connection to each producer
+            self._sub_tasks = [
+                self._loop.create_task(self._handle_producer_connection(producer, rpc_lock))
+                for producer in self.producers
+            ]
+
+            await asyncio.gather(*self._sub_tasks)
+        except asyncio.CancelledError:
+            pass
+        finally:
+            # Stop the loop once we are done
+            self._loop.stop()
+
+    async def _handle_producer_connection(self, producer, lock):
+        """
+        Main connection loop for the consumer
+        """
+        try:
+            while True:
+                try:
+                    url, token = producer['url'], producer['ws_token']
+                    ws_url = f"{url}?token={token}"
+
+                    async with websockets.connect(ws_url) as ws:
+                        logger.info("Connection successful")
+                        channel = WebSocketChannel(ws)
+
+                        # Tell the producer we only want these topics
+                        # Should always be the first thing we send
+                        await channel.send(
+                            self.compose_consumer_request(RPCRequestType.SUBSCRIBE, self.topics)
+                        )
+
+                        # Now receive data, if none is within the time limit, ping
+                        while True:
+                            try:
+                                message = await asyncio.wait_for(
+                                    channel.recv(),
+                                    timeout=5
```

Comment:
isn't pinging every 5s a bit excessive?

Reply:
Oh I must have missed this, it should reference `self.ping_timeout`

Reply:
We can change the default to 30 seconds

Reply:
configurable is even better. we can change the default shoudl we notice problems or excessive traffic.